consensus/core/src/dag_state.rs

```diff
@@ -257,6 +257,48 @@ impl DagState {
         Ok(blocks)
     }
 
+    /// Returns the last block proposed per authority with round <= `before_round`. The method will look
+    /// into cache first, otherwise will fall back in store. The method guarantees to return a result for
+    /// each authority, even if that's genesis block.
+    pub(crate) fn get_last_block_per_authority(
+        &self,
+        before_round: Round,
+    ) -> ConsensusResult<Vec<VerifiedBlock>> {
+        let mut blocks = vec![None; self.context.committee.size()];
+
+        for (block_ref, block) in self.recent_blocks.range((
+            Included(BlockRef::new(0, AuthorityIndex::ZERO, BlockDigest::MIN)),
```

Comment:
This iterates over almost all recent blocks. It seems more efficient to check the latest block before limit round per authority in `cached_refs`.

Reply:
ok I can't even think _why I_ did this yesterday this way - it's obviously not computationally frugal. I'll refactor, thanks for pointing out